### PR TITLE
Add filename pattern to reader associations to preference dialog

### LINKED ISSUE #4146: UI for direct assignment of file extension to reader plugin
## 🚀 Feature

Since npe2 and #3799 we can now associate readers to file extensions when dragging and dropping files or via the command line. We can also view these associations in preferences (and remove them if necessary). 

However, we still don't have a GUI mechanism for creating associations outside of opening files. #3502 was the first pass at creating such a UI, though it focused on both readers and writers.

napari

I think writers are more complex because we also have to consider what layers are being written - e.g. `npe2` plugins might declare that they take `[image+, labels{2}, ...]`, which doesn't seem easy to capture via the GUI. The writer issue has also been mostly addressed by allowing the user to select writers from the file dialog.

## Motivation

Users should be able to add extension to reader associations outside of dragging and dropping a file.

## Pitch

I think rather than allowing the user to type extensions into the table (as in the gif above) we should leverage npe2 plugins' extension declarations. Essentially, I think we should have a two column table (Extension, Reader). User clicks on a new row in 
the extensions column and selects an extension from a populated list of the union of all plugins' declared extensions. Once the extension is selected, the user can click a dropdown in the Reader column and select from a list of reader plugins that has been filtered down to only those that accept this extension. Original `napari-plugin-engine` plugins are shown for all extensions. Maybe @isabela-pf could mock up something like this? 

## Alternatives

Alternatives would be allowing the user to type in extensions and select plugins and maybe flag if the chosen coupling wouldn't work? 

Totally open to hearing other people's ideas, UX is not my strong point.

## Additional context

More context can be found in #3502.

### PR DESCRIPTION
**Pull request by DragaDoncila**
# Description
Prior to this PR filename associations could only be made in the console and were limited to file extensions.  Initial instinct in the community meeting on whether we should allow saving *just* extensions or the full range of patterns supported by `fnmatch` was that since npe2 readers support full filename matching, so should the dialog. This PR therefore adds the ability to associate filename patterns with compatible readers in the preference dialog, and fixes the preferred reader logic to also make use of the broader filename pattern matching.

https://user-images.githubusercontent.com/17995243/166202234-6e19da97-e8c5-4db6-b6a2-b9736d1a7ad6.mov

@junxini I know you wanted to have a look at this. @isabela-pf would love your feedback also on the UX part.

## Notes
- For simplicity, the reader dialog still only supports a checkbox to save a preference for the file extension (if a file extension is present). More complex pattern association is available through the preference dialog only.
- I've tried to move away from referring to `extensions` in this context in the code, but the underlying model setting is still `extension2reader` - I'm not sure what it would take to change that setting or whether it's worth it
- Since we don't yet support association or selection of individual reader contributions within a single plugin, a plugin is deemed compatible by the filter when any of the accepted patterns of its readers match the user's pattern.

## Outstanding Questions
1. ~~At the moment this wouldn't be backwards compatible for users who had previously stored a preference as e.g. `.tif` rather than `*.tif`. Is there a way to make sure we fix existing preferences?~~ We've addressed this with a setting migration.
2. Currently we don't check whether two settings might conflict. We need some UX design about what to do in that situation. This will be investigated as part of work already scheduled to improve the file opening behavior and reader plugin discovery in the viewer.

## Type of change
- [x] New feature (non-breaking change which adds functionality)
- ~~[x] Breaking change (fix or feature that would cause existing functionality to not work as expected) - (ish.. see Question 1 above)~~ See Question 1 above

# References
Closes #4146.

# How has this been tested?

- [x] added tests for new behavior

## Final checklist:
- [ ] My PR is the minimum possible work for the desired functionality
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] If I included new strings, I have used `trans.` to make them localizable.
      For more information see our [translations guide](https://napari.org/docs/dev/guides/translations.html).

### CODE REVIEW

**Comment by junxini** [comment]
This looks great Draga! 

> Currently we don't check whether two settings might conflict.

Could you help give an example of 2 reader preference settings that might conflict?

**Comment by DragaDoncila** [comment]
@junxini yes, I can give an example. 

Let's say I have `napari_tifffile` installed, which accepts all filename patterns. I choose to associate `*.tif` so it opens all TIFFs.

I also have a different more specific TIFF reader `Draga's Test Reader`, that only takes in tiff files within a folder named a specific way (I've named it `*/my-specific-folder/*.tif` but maybe it would be a folder name that defines what's inside e.g. `*/training-datasets/Fluo-C2DL-Huh7/*.tif ` (this is an example from [the tracking challenge](http://celltrackingchallenge.net/2d-datasets/). 

Anything that matches ``*/my-specific-folder/*.tif`` also matches `*.tif` because the latter is a more lenient constraint - just give me anything that ends in `.tif` essentially. This means you can get a different plugin claiming the file depending on what order you saved your preferences in.

Getting more technical, I return the first reader whose associated pattern matches the filename we're trying to open. This makes the preference list act like an access control list. I'm not saying this is the best approach - we probably want to at least highlight conflicts if they occur.

I think this is unlikely to be a frequent issue for users in the time it takes us to explore better behavior as part of the ABRF work. I'm more concerned about preferences for extensions now needing a `*` in front, because of the nature of `fnmatch` - we should be able to fix existing user settings, I'm just not sure where that would go - maybe @tlambert03 knows how to "patch" a user's settings in this case.

Let me know if you'd like me to elaborate on anything!

**Comment by junxini** [comment]
Thank you for explaining Draga!

**Comment by isabela-pf** [comment]
@DragaDoncila thanks for the tag! This is an interesting detail to work out!

Overall, this interactions meets what I'd expect of similar presets. I have clarifying questions to make sure I'm not missing anything.

Replies to things you wrote: 

> For simplicity, the reader dialog still only supports a checkbox to save a preference for the file extension (if a file extension is present). More complex pattern association is available through the preference dialog only.

Do preferences saved via dialog appear in the Preferences list afterwards? I would want them to, if you haven't already done that.

> Since we don't yet support association or selection of individual reader contributions within a single plugin, a plugin is deemed compatible by the filter when any of the accepted patterns of its readers match the user's pattern.

That makes sense! Is there any kind of feedback for if a user wants to set a preset for something that plugin does not accept? 

> At the moment this wouldn't be backwards compatible for users who had previously stored a preference as e.g. .tif rather than *.tif. Is there a way to make sure we fix existing preferences?

I think this is an implementation question, but if we can't make this update automatically there may be something we can do. I'm imagining this one be a one-time "you'll lose your preferences with this update" kind of situation. Not ideal, but workable if we can't make something else work.

> Currently we don't check whether two settings might conflict. We need some UX design about what to do in that situation. This will be investigated as part of work already scheduled to improve the file opening behavior and reader plugin discovery in the viewer.

That's my cue! Thanks for the above example of a conflict, too. When you say `"I return the first reader whose associated pattern matches the filename we're trying to open."` is the first alphabetical, or more strict before more lenient, or something else?

I also agree it may be good to let users choose which pattern they want to follow when there are conflicting presets. I may not understand this deeply enough, but since opening a file with an extension that has conflicting presets would prevent the opening of said file until a decision is made, this may be the perfect use for a dialog asking users to choose. Does that seem fitting?

Other questions:
- Patterns can't be edited, correct? Just deleted and re-added? I think that's okay since these aren't complicated inputs.
- What happens if the list gets longer than the dialog height? Does it add a scrollbar to this section of Preferences?
- How do the patterns act if they are longer than the section of the table they are given? For example, do they truncate the end of a file path or plugin name with a `…`?

**Comment by DragaDoncila** [comment]
@isabela-pf thanks so much for the thorough review and great questions. I've made a few changes based on your questions/suggestions.

> Do preferences saved via dialog appear in the Preferences list afterwards? I would want them to, if you haven't already done that.

Yep, they appear in the list just like any other preference.

> Is there any kind of feedback for if a user wants to set a preset for something that plugin does not accept?

You can't set a preference for something a plugin doesn't accept. In the worst case, no plugins can read your filename pattern and the dropdown list is empty. I've modified it to say that there's nothing available (so the user knows it's not an error).

Screen Shot 2022-05-08 at 5 33 22 PM

We could provide a way for a user to just say "no I don't care if it's not compatible, assign the preference anyway"? But tbh I don't really see much use for that.

> I'm imagining this one be a one-time "you'll lose your preferences with this update" kind of situation. Not ideal, but workable if we can't make something else work.

I've figured out how to update existing users' settings to be in line with the new approach, so this is no longer an issue. Preferences will be "silently" updated, with no change in behavior for users.

> When you say "I return the first reader whose associated pattern matches the filename we're trying to open." is the first alphabetical, or more strict before more lenient, or something else?

The "first" will be the first preference that was added by the user - this order is reflected in the table. The top row was added first, the last row was added last. This can be useful because users can go from more specific to less specific patterns, but I haven't documented/suggested this in instructions, since we'll be reviewing conflicts more closely as part of related work.

>  I may not understand this deeply enough, but since opening a file with an extension that has conflicting presets would prevent the opening of said file until a decision is made, this may be the perfect use for a dialog asking users to choose. Does that seem fitting?

I may be wrong, but I actually don't think it's possible for a conflict to result in a file not opening. In the worst case, if your more general `*.tif` preference was above your more specific `my-specific-folder/*.tif` preference, it would be opened with the more general plugin.

> Patterns can't be edited, correct? Just deleted and re-added?

They actually can - if a user types a pattern that is already in the preference list, the preference is swapped out with the new reader choice. We may want to warn in this case, but I don't think we should stop them. This decision can be part of the design audit that will soon be done as part of the work to support the ABRF file opening workflow. However if anyone thinks it's critical, we can of course tweak it now.

> What happens if the list gets longer than the dialog height?

> How do the patterns act if they are longer than the section of the table they are given?

There is a scrollbar for very long preference lists, and if the pattern is very long a `...` is added. You can see both in the screenshot below.

Screen Shot 2022-05-07 at 9 59 06 PM

It's worth noting that the preference list will soon have more vertical room, as we will remove the plugin call order preferences.

**Comment by isabela-pf** [comment]
Thank *you* for such a thorough response. I can call out specifics, but I think I agree with everything you said. You have my approval! (Let me know if you want me to actually hit the big, green Approval button.)

---

> You can't set a preference for something a plugin doesn't accept. In the worst case, no plugins can read your filename pattern and the dropdown list is empty. I've modified it to say that there's nothing available (so the user knows it's not an error).

Fantastic!

> We could provide a way for a user to just say "no I don't care if it's not compatible, assign the preference anyway"? But tbh I don't really see much use for that.

I agree there doesn't seem to be a use for that. It seems counter to a good experience to let people set up something that just…will fail? It also sounds like this would be easy to add if someone comes up with a reason later.

> I've figured out how to update existing users' settings to be in line with the new approach, so this is no longer an issue. Preferences will be "silently" updated, with no change in behavior for users.

I bow down to your immeasurable skill and power. ✨ 

> The "first" will be the first preference that was added by the user - this order is reflected in the table.

That makes sense! Out of curiosity, can users reorder this?

> I may be wrong, but I actually don't think it's possible for a conflict to result in a file not opening. In the worst case, if your more general *.tif preference was above your more specific my-specific-folder/*.tif preference, it would be opened with the more general plugin.

Sounds good to me! I love to hear things won't totally fail.

> There is a scrollbar for very long preference lists, and if the pattern is very long a ... is added. You can see both in the screenshot below.

Thanks for showing me. Looks good.

File: napari/_qt/widgets/qt_extension2reader.py [hunk 1]
```diff
@@ -20,94 +30,248 @@ class Extension2ReaderTable(QWidget):
...
         self._populate_table()
 
+        instructions = QLabel(
+            trans._(
+                'Enter a filename pattern to associate with a reader e.g. "*.tif" for all TIFF files.'
+                + 'Available readers will be filtered to those compatible with your pattern. Hover over a reader to see what patterns it accepts.'
+                + '\n\nPreference saving for folder readers is not supported, so these readers are not shown.'
+                + '\n\nFor documentation on valid filename patterns, see https://docs.python.org/3/library/fnmatch.html'
+            )
+        )
```
Comment (jni) [inline napari/_qt/widgets/qt_extension2reader.py]: I think translations need to be on a single line, so each line needs to be `trans._()`d separately and then concatenated. (Or you could do a big-ass line but I like concatenating the translations better.)

**Review by jni — APPROVED** [review]
So many tests! :heart_eyes:

**Comment by jni** [comment]
@napari/core-devs I intend to merge this shortly to get it in 0.4.16rc2, if anyone has objections let me know. Final release won't be for 1-2 days so it can be reverted if issues come up.

**Review by tlambert03 — APPROVED** [review]
Haven't looked at the tests... but am made all the more comfortable by their presence :)

This looks good to move forward!  Thanks @DragaDoncila

**Pull request merged**